PR: Add code folding to the Editor

rlaverde opened the pull request:
Fixes #706 

----

Progress:
- Add pyqode code `decoration.py` `folding.py` `utils.py` `codefolding.py` `managers/decoration.py`
- Register FoldingPanel to codeeditor (at the moment It's displayed as an empty panel)

TODO:
- [x] Reconnect Signals
- [x] Fix comment out editor.add.menu()
- [x] Delete inner lines if folding line is deleted
- [x] Implement pyqode SyntaxHighlighter.highlightBlock functionality that call FoldDetector.process_block, in each block
- [x] Fix error with Syntax higtlighter and multiple line comments
- [x] Fix error with mouse  event (mouse position is a little wrong)
- [x] Change docstrings to make reference to spyder classes
- [x] Rebase it after #3778 
- [x] Add pyqode copyright and license and links to source files
- [x] Reword, reorder and squash some commits
- [x] Test that everything is working

~~Somethings that I'm not sure if should be do it:~~
~~- Remove decoration.py and implement It's functionality in CodeEditor~~ I think is ok to leave it
~~- Remove utils.py and use CodeEditor functionality (and implement missing functionalities)~~ I think is ok to leave it too

Things that I'll address in other PRs
- Move some codeeditor functions to TextHelper, and delete redundant functions
- Inners foldings should be preserved when folding unfolding their parent
- Add shortcuts and menu for codefolding actions
- Some style improvements (only show arrows when mouseover, different colors, only decorate selected block, ~~use different icons for folding arrows~~)
- Implement others fold detectors (apart of the indentation fold detector)

ccordoba12 commented:
Great work @rlaverde!

Could you post a screenshot for this new feature as part of the description? Thanks!

rlaverde commented:
@ccordoba12 at the current state it's not possible, It just show an empty panel, I think that after implemented the missing part in SyntaxHighlighter, it will work :)

rlaverde commented:
There are ~~a lot~~ some things missing, but It's working now :smile: 
![code_folding](https://cloud.githubusercontent.com/assets/2024217/21367454/3ef2743e-c6cc-11e6-902b-7554ca0e7aa8.png)

ccordoba12 commented:
This is really cool!! Thanks a lot @rlaverde!!

How folding is computed? Using regexs?

rlaverde commented:
> How folding is computed? Using regexs?

Using Indentation, although It could be extended to use other methods
https://github.com/spyder-ide/spyder/pull/3833/files#diff-6ceb5fb1fd412eaffd5a62296d5c688bR138

ccordoba12 commented:
> Using Indentation, although It could be extended to use other methods

Ok, that's fine for Python, but we definitely need to add other methods for languages like C/C++ and Fortran ;-)

In any case this is a hell of a improvement!! Good job :-)

rlaverde commented:
@goanpeca @ccordoba12  This is ready, only missing a rebase with master after #3778 merge

goanpeca commented:
Woot! so baby jesus is getting code foldng after all :laughing:

ccordoba12 commented:
Great work @rlaverde! Really nice Christmas present for Spyder!!

I have one question: is this only applied to Python files (for now)?

ccordoba12 commented:
One important suggestion before reviewing: please use FontAwesome icons for the folding marks, and not just the PNG's you provided. That's very important to preserve not only consistency but also our ability to render correctly in high DPI screens.

goanpeca commented:
>One important suggestion before reviewing: please use FontAwesome icons for the folding marks, and not just the PNG's you provided. That's very important to preserve not only consistency but also our ability to render correctly in high DPI screens.

@rlaverde yes, could we use those icons better :-) ?

ccordoba12 commented:
Png's are fine for the Spyder 2 theme, but we also need to use FontAwesome icons :-)

goanpeca commented:
@rlaverde could you rebase now that panels are merged?

rlaverde commented:
> One important suggestion before reviewing: please use FontAwesome icons for the folding marks, and not just the PNG's you provided. That's very important to preserve not only consistency but also our ability to render correctly in high DPI screens.

Ready, I deleted old pngs and use caret font awesome icons. (I squashed the commit)

> could you rebase now that panels are merged?

I just rebase it with master :smile:

goanpeca commented:
@ccordoba12 are we planning on having pngs with Spyder 4.0?

I would say that on 4.0 we should only support ttf fonts via qtawesome?

ccordoba12 commented:
> I would say that on 4.0 we should only support ttf fonts via qtawesome?

I agree, but until we do that we're still supporting the Spyder 2 theme :-)

goanpeca commented:
>I agree, but until we do that we're still supporting the Spyder 2 theme :-)

But this is on Master... so why should we support it on master?

ccordoba12 commented on spyder/api/utils.py:
```diff
@@ -0,0 +1,1096 @@
+# -*- coding: utf-8 -*-
```
I don't think this should be exposed as part of our API because it's composed purely of helper functions.

Please move the contents of this file to `utils`, finding a right place for each function in our current files. If no place can be found, please create a new `utils/folding.py` file.

ccordoba12 replied:
Or maybe you need to create a `utils/folding` module ;-)

rlaverde replied:
this utils functions aren't specific for codefolding, I'll name it `utils/editor.py`

ccordoba12 commented on spyder/api/utils.py:
```diff
@@ -0,0 +1,1096 @@
+from qtpy.QtWidgets import QApplication
+
+
+class memoized(object):
```
We already have a function in `utils` to do memoizations. Please try to merge this function with that one.

rlaverde replied:
This function isn't used, I'll delete it

ccordoba12 commented on spyder/widgets/sourcecode/api/decorations.py:
```diff
@@ -0,0 +1,65 @@
+# -*- coding: utf-8 -*-
```
Could we add this to `sourcecode/__init__.py`? Or maybe in `sourcecode/manager.py`?

I mean, it's a bit strange to have an `api` module that's not exposed in `spyder.api` :-)

rlaverde replied:
It was @goanpeca idea to create another api just for the editor, I'm not sure where to put this, maybe `sourcecode/manager.py` will be better, I don't like the pidea opf put code in `__init__.py`

goanpeca replied:
@ccordoba12 the idea is that complex widgets will have some sort of api, this is the case of the editor. It would be very weird to decouple the api of the code editor and put it in a "global" spyder.api folder. The idea is for widgets to be self contained as all other spyder widgets.

ccordoba12 replied:
Ok, good idea. I haven't thought about it :-)

ccordoba12 commented on spyder/api/decoration.py:
```diff
@@ -0,0 +1,205 @@
+
+class TextDecoration(QTextEdit.ExtraSelection):
+    """
+    Helper class to quickly create a text decoration.
```
Since this file only contains a helper class, it should be moved to `utils` :-)

rlaverde replied:
I think It should be moved to the same place that `widgets/sourcecode/api/decorations.py`

ccordoba12 replied:
Ok

ccordoba12 commented:
I did a very quick review just to suggest how to better organize the new code :-)

ccordoba12 commented:
> But this is on Master... so why should we support it on master?

@stonebig likes to pack our betas in WinPython. However, we could remove the Spyder 2 theme in beta1 to avoid future problems with that ;-)

Please open an issue an assign it to one of our Junior devs.

goanpeca commented:
will do

durack1 commented:
@ccordoba12 @goanpeca @rlaverde excellent!! Looking forward to giving this a run.. When are you planning on releasing the v4.0 betas through conda..?

ccordoba12 commented:
> When are you planning on releasing the v4.0 betas through conda..?

I think we'll release the first beta during the summer (July/August), hoping for a final release by the end of the year.The thing is we're working to define a public API for the project, and that takes careful planning and move a lot of code around.

rlaverde commented:
@ccordoba12 I made the changes that you suggested, could you review this PR again.

goanpeca commented:
@rlaverde there seem to be some conflicts can you rebase or merge with master?

ccordoba12 commented on spyder/api/folding.py:
```diff
@@ -0,0 +1,365 @@
+        raise NotImplementedError
+
+
+class IndentFoldDetector(FoldDetector):
```
I think from this point to the end of the file doesn't belong here, i.e. we shouldn't expose it as part of our API because it includes particular implementations of `FoldDetector` (and `FoldScope` is a utility class).

So I propose to move them to `widgets/sourcecode/folding.py`

----

In my opinion, the big picture for `spyder.api` is to put here *only* what's needed to extend Spyder. So it shouldn't contain implementations nor utilities, just the basic building classes and modules that we give people create their own plugins.

goanpeca replied:
So I think it is better that every widget has its own internal api (also in an api file maybe?)

And on spyder.api we only expose what we want people to have access to. This can be the whole widget api or maybe a subset. In this case we can put evrything related to the editor and the source code editor in `spyder.api.editor.py`

ccordoba12 replied:
I agree with that, but again `IndentFoldDetector` and `CharFoldDetector` are implementations, so they don't belong to `CodeEditor` API.

goanpeca replied:
Yep agreed :-)

ccordoba12 commented on spyder/api/folding.py:
```diff
@@ -0,0 +1,365 @@
+
+    To use a FoldDetector, simply set it on a syntax_highlighter::
+
+        editor.syntax_highlighter.fold_detector = my_fold_detector
```
Please add to this docstring this

    Some implementations of fold detectors can be found in
    ``spyder/widgets/sourcecode/folding.py``.

ccordoba12 commented on spyder/api/folding.py:
```diff
@@ -0,0 +1,365 @@
+    :param editor: CodeEditor instance.
+    :param file: file handle where the tree will be printed. Default is stdout.
+    :param print_blocks: True to print all blocks, False to only print blocks
+        that are fold triggers
```
I think it's fine to leave this here so people can easily debug their detectors.

ccordoba12 commented on spyder/api/folding.py:
```diff
@@ -0,0 +1,365 @@
+        return TextBlockHelper.get_fold_lvl(prev_block)
+
+
+class FoldScope(object):
```
When moving this `widgets/sourcecode/folding.py` please put this class at the beginning of that file.

ccordoba12 commented on spyder/widgets/sourcecode/api/decoration.py:
```diff
@@ -0,0 +1,205 @@
+# -*- coding: utf-8 -*-
```
You have `sourcecode/api/decoration.py` and `sourcecode/api/decorations.py`, which is confusing.

Since this only contains a helper class, we could move it to `sourcecode/utils/decoration.py` or something like that.

ccordoba12 commented:
@rlaverde, I left a second round of comments. Besides, this PR needs a rebase/merge with master to solve conflicts.

rlaverde commented:
@ccordoba12 I rebase it, and address your comments.

I also add a `spyder.api.editor` to expose some parts of internal `sourcecode.api`

goanpeca commented:
awesome :-)

goanpeca commented:
@rlaverde there are some conflicts

goanpeca commented:
@ccordoba12 unless you have more comments this is ready to be merged

ccordoba12 commented:
Ok, let's go with it so that @rlaverde can keep enhancing it on master.

This is a huge improvement!!! Thanks @rlaverde for working on it :-)
